### DISCUSSION TITLE
Separate "probability expression" syntax

### DISCUSSION THREAD

**Discussion by phipsgabler** [TuringLang/AbstractPPL.jl#13] [Ideas]
I've been contemplating whether something like StatsModels [formula syntax](https://juliastats.org/StatsModels.jl/stable/internals/#Internals-and-extending-the-formula-DSL), but for probability expressions, might be useful to abstract out as well.

For example, this parses quite OK:

```julia
julia> :(f(first, @P(X | (; Y = 2)), model)) |> Meta.show_sexpr
(:call, :f, :first, (:macrocall, Symbol("@P"), :(#= REPL[13]:1 =#), (:call, :|, :X, (:tuple, (:parameters, (:kw, :Y, 2))))), :model)
```

There are currently two places where something similar is used:

1. In DynamicPPL's [`prob` string macros](https://turing.ml/dev/docs/using-turing/guide#querying-probabilities-from-model-or-chain) for probability queries:

    ```julia
    prob"s = 1.0, m = 1.0 | model = gdemo, x = nothing, y = nothing"
    ```

    There, the string is really just split at the `|`, and then `Meta.parse`d.  Separating out probability expressions would remove some awkwardness of "conditioned variables" which are not random variables, like the `model`.
2. In (future) Soss, you should be able to condition models: https://github.com/cscherrer/Soss.jl/pull/216. Not quite the same thing, but going into a similar direction.

Perhaps other things, like prior/likelihood queries and transformations against models, can be subsumed under this, too.

CC @yebai @mohamed82008 @cscherrer

**Comment by cscherrer** [comment]
I like the `@P` macro idea. What would `first` be in this example?

Soss has `|` notation in dev already. I've been using it for a while, and I think it's working well. I need to get this into master; it's just going to be a giant merge at some point. Refactoring is hard.

I was also wondering about `|` notation for MeasureTheory, for restricting a measure to a subset. This is really what happens in any PPL anyway. We never compute the normalizing factor, so the `logdensity` or `logpdf` is really either not a probability measure, or it's no longer wrt Lebesgue measure. Might as well recognize this :)

**Reply by phipsgabler** [reply]
`first` is nothing relevant, just a leftover from when I tested how the `@P` call parses within a function call.

Although I had been thinking of notation like `query(model, @P ...)`.
